[ADD] Add new addon event_project

Endika opened the pull request:
This module allows you to create retro planning for managing your events.
To use this module, you need to:

-First step, you create a project template.
-Second step, you indicate in to task of template the previous days to begin task.
-And Last step, you create a new event and select project template.

Automatically, will be create a new related project with event.

Commit:
[ADD] Add new addon event_project

Commit:
[ADD] lang

Commit:
[FIX]

Commit:
[UPDATE] README

BBerastegui commented:
:+1:

JavierIniesta commented:
:+1:

IgorPastor commented:
:+1:

dreispt commented on event_project/models/event_event.py:
```diff
@@ -0,0 +1,105 @@
+        comodel_name='project.project', string='Related project',
+        readonly=True)
+
+    def reorganize_project(self, project, event, date_begin=None):
```
For better separation of concerns, shouldn't this logic be on project.project, and this method only make a call on it?

dreispt commented on event_project/README.rst:
```diff
@@ -0,0 +1,52 @@
+Module name
+===========
+
+This module allows you to create retro planning for managing your events.
```
I wish we had a better description.
<s>Shouldn't "retro planning" be a generic Project feature reused for this use case?</s> (I can see that it's quite specific...)

Commit:
[UPDATE] README and add project_project model.

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,55 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Module name
```
Put the real name of the module

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,55 @@
+And if the event date start modify, recalculate tasks dates.
+
+
+Installation
```
Don't put this section as all are standard modules. The section is reserved when you have any dependency.

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,55 @@
+
+To use this module, you need to:
+
+-First step, you create a project template.
```
Remove `you`

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,55 @@
+To use this module, you need to:
+
+-First step, you create a project template.
+-Second step, you indicate in to task of template the previous days to begin task.
```
Remove `you` (imperative goes without noun or pronoun)

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,55 @@
+
+-First step, you create a project template.
+-Second step, you indicate in to task of template the previous days to begin task.
+-And Last step, you create a new event and select project template.
```
Remove you

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,55 @@
+-First step, you create a project template.
+-Second step, you indicate in to task of template the previous days to begin task.
+-And Last step, you create a new event and select project template.
+Automatically, will be create a new related project with event.
```
For a better formatting, put all these point with `*` preceding them

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,55 @@
+Module name
+===========
+
+This module allows you to create events with assigned a project.
```
Rewrite this with:
"This module allows you to assign a project template to an event, creating the corresponding project for the event that ends on the event start date, and plan the project tasks backwards according a new field "Event anticipation days".

pedrobaeza replied:
Add at the end:
"If the event date is changed, then all the pending tasks are retro-planned again according the new date."

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,55 @@
+Contributors
+------------
+
+* Antiun Ingenieria S.L. (Madrid, Spain, http://www.antiun.com)
```
Companies are not contributors, unless it is a generic reference (which is not the case).

pedrobaeza commented on event_project/i18n/ca.po:
```diff
@@ -0,0 +1,46 @@
+
+#. module: event_project
+#: field:project.task,previous_day:0
+msgid "Previous day"
```
Call it "Event anticipation days"/"Días de anticipación del evento"

pedrobaeza commented on event_project/i18n/ca.po:
```diff
@@ -0,0 +1,46 @@
+#. module: event_project
+#: field:event.event,project_related:0
+msgid "Related project"
+msgstr "Projecto relacionado"
```
s/Projecto/Proyecto

pedrobaeza commented on event_project/i18n/ca.po:
```diff
@@ -0,0 +1,46 @@
+#. module: event_project
+#: field:event.event,project_template:0
+msgid "Template project"
+msgstr "Plantilla de projecto"
```
s/projecto/proyecto

pedrobaeza commented on event_project/i18n/es.po:
```diff
@@ -0,0 +1,46 @@
+
+#. module: event_project
+#: field:project.task,previous_day:0
+msgid "Previous day"
```
Call it "Event anticipation days"

pedrobaeza commented on event_project/i18n/eu.po:
```diff
@@ -0,0 +1,46 @@
+# Translation of Odoo Server.
```
Why creating this file if the translation contents are spanish?

Endika replied:
Ok, now I haven't time to translate, so will remove files.

pedrobaeza commented on event_project/i18n/ca.po:
```diff
@@ -0,0 +1,46 @@
+# Translation of Odoo Server.
```
This is not catalonian, it's spanish, isn't it?

Endika replied:
Ok, now I haven't time to translate, so will remove files.

pedrobaeza commented on event_project/i18n/gl.po:
```diff
@@ -0,0 +1,46 @@
+# Translation of Odoo Server.
```
The same comment

Endika replied:
Ok, now I haven't time to translate, so will remove files.

pedrobaeza commented on event_project/models/event_event.py:
```diff
@@ -0,0 +1,85 @@
+    project_template = fields.Many2one(
+        comodel_name='project.project', string='Template project',
+        domain="[('state', '=', 'template')]")
+    project_related = fields.Many2one(
```
Call it only `project`. The many2one field indicates that it's the "related" project :wink:

pedrobaeza commented on event_project/models/event_event.py:
```diff
@@ -0,0 +1,85 @@
+        project_template = None
+        project_related = None
+        date_begin = None
+
```
Don't put extra blank lines, because make more difficult to read Python code.

Endika replied:
This extra blank lines is for validate with pep8,, pep257 and others

pedrobaeza replied:
No, it doesn't. Blank lines is only for splitting methods or classes.

pedrobaeza commented on event_project/models/project_task.py:
```diff
@@ -0,0 +1,33 @@
+class ProjectTask(models.Model):
+    _inherit = 'project.task'
+
+    previous_day = fields.Integer(
```
Call it `anticipation_days` to better indicate the purpose, and the label can be "Event anticipation days".

pedrobaeza commented on event_project/views/event_event_view.xml:
```diff
@@ -0,0 +1,18 @@
+
+        <field name="date_tz" position="after">
+            <field name="project_related" attrs="{'invisible':[('project_template','=',False)]}"/>
+            <field name="project_template" attrs="{'readonly':[('project_related','!=',False)],'invisible':[('project_related','!=',False)]}"/>
```
Don't make it invisible. You're already preventing its change with readonly flag.

pedrobaeza commented on event_project/models/event_event.py:
```diff
@@ -0,0 +1,85 @@
+        if (event.project_template and not event.project_related):
+            project_obj = self.env['project.project']
+            event.project_template.duplicate_template()
+            obj_ids = project_obj.search([])
```
Instead of this hackish method (that fails if the order of the model changes to another key different from id), you can get the result of the previous method and test `res_id` key:

```
result = event.project_template.duplicate_template()
event.project_related = result['res_id']
```

pedrobaeza commented on event_project/models/event_event.py:
```diff
@@ -0,0 +1,85 @@
+        project_related = None
+        date_begin = None
+
+        if ((self.project_template or
```
Here you only need to test 2 things:
- `if vals.get('project_template')`: This is for assigning a template and make the same as when you create an event. You must extract the logic to one method for it to avoid duplicate code.
- `if vals.get('start_date')`: This means that the event date has changed, so you have to call the function that changes tasks dates.

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,55 @@
+                date_begin = datetime.strptime(
+                    event.date_begin, "%Y-%m-%d %H:%M:%S")
+
+            if project_task.previous_day > 0:
```
You should put `if project_task.previous_day:` and allow also negative dates for tasks after the event (for example, a thanks giving mailing after the event)

pedrobaeza replied:
Thinking twice, you don't need this condition, when project_task.previous_day == 0, the timedelta operation returns date_begin, so everything is right.

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,55 @@
+            if project_task.previous_day > 0:
+                pdays = int(project_task.previous_day) * -1
+                date_start = (date_begin + timedelta(days=pdays)).strftime(
+                    "%Y-%m-%d %H:%M:%S")
```
Use `fields.Datetime.to_string()` instead of strftime.

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,55 @@
+                    date_begin = datetime.strptime(
+                        date_begin, "%Y-%m-%d %H:%M:%S")
+            else:
+                date_begin = datetime.strptime(
```
Use `fields.Datetime.from_string()` instead of strptime.

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,55 @@
+        for obj_id in obj_ids:
+            project_task = project_task_obj.browse(int(obj_id))
+            if date_begin:
+                if type(date_begin) is str:
```
Remove this check. It responsability of the programmer to pass the correct type.

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,55 @@
+            project_task = project_task_obj.browse(int(obj_id))
+            if date_begin:
+                if type(date_begin) is str:
+                    date_begin = datetime.strptime(
```
Use `fields.Datetime.from_string()` instead of strptime.

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,55 @@
+                    event.date_begin, "%Y-%m-%d %H:%M:%S")
+
+            if project_task.previous_day > 0:
+                pdays = int(project_task.previous_day) * -1
```
Why multiply by -1? You can just put the sign `-` ahead, or better yet, make the substraction in the next instruction instead the sum: `date_start = (date_begin - timedelta(days=pdays...`

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,55 @@
+                date_start = (date_begin + timedelta(days=pdays)).strftime(
+                    "%Y-%m-%d %H:%M:%S")
+                project_task.write({'date_start': str(date_start)})
+            else:
```
Remove this else part as we remove the condition.

pedrobaeza commented on event_project/views/project_task_view.xml:
```diff
@@ -0,0 +1,17 @@
+      <field name="arch" type="xml">
+
+        <field name="planned_hours" position="after">
+            <field name="previous_day"/>
```
Put the new name.

pedrobaeza commented:
For the icon, please make a collage (or ask Dani to make it) of the event and the project standard icon instead of this generic one.

Commit:
[UPDATE] multiple update

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,44 @@
+* First step, create a project template.
+* Second step, indicate in to task of template the previous days to begin task.
+* And Last step, create a new event and select project template.
+Automatically, will be create a new related project with event.
```
- It will then create automatically a new project related with the event.

pedrobaeza commented on event_project/README.rst:
```diff
@@ -0,0 +1,44 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :alt: License: AGPL-3
+
+Event project
```
Plan events through projects

pedrobaeza commented on event_project/models/event_event.py:
```diff
@@ -0,0 +1,81 @@
+        project_template = None
+        project = None
+        date_begin = None
+
```
Remove intermediate blank lines inside a method

pedrobaeza commented on event_project/models/event_event.py:
```diff
@@ -0,0 +1,81 @@
+        event = super(EventEvent, self).create(vals)
+        if (event.project_template and not event.project):
+            project_obj = self.env['project.project']
+            result = event.project_template.duplicate_template()
```
Extract this code to a method and call it from here and from write method (you're doing the same twice).

Endika replied:
I do not understand that you want to extract.
This line:

> result = event.project_template.duplicate_template()

pedrobaeza replied:
That line and the assignation to the event. Take in note you have made in the create the tasks after the super call, and in write before. Do both in the same place.

pedrobaeza commented on event_project/models/event_event.py:
```diff
@@ -0,0 +1,81 @@
+        project = None
+        date_begin = None
+
+        if ((self.project_template or
```
You don't have to check this complex expressions. I told you in previous comments what to check (https://github.com/OCA/crm/pull/26#discussion_r30159859)

Endika replied:
I put this:

> if ((self.project_template or vals.get('project_template')) and not self.project)

I need check both cases self.project_template and vals['project_template']

pedrobaeza replied:
No, you don't: vals dictionary on write only passes values that have changed, so testing only `vals.get('project_template')` it's enough to know if you have to create the project.

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,45 @@
+            project_task = project_task_obj.browse(int(obj_id))
+            if not date_begin:
+                date_begin = fields.Datetime.from_string(event.date_begin)
+
```
Remove this empty line

pedrobaeza commented on event_project/models/project_project.py:
```diff
@@ -0,0 +1,45 @@
+    def reorganize_project(self, event, date_begin=None):
+        project_task_obj = self.env['project.task']
+        self.write({'name': event.name})
+        obj_ids = project_task_obj.search([('project_id', '=', self.id)])
```
This query return a recordset, not a list of ids, so you don't need to make again the browse.

pedrobaeza commented on event_project/models/project_task.py:
```diff
@@ -0,0 +1,33 @@
+    _inherit = 'project.task'
+
+    anticipation_days = fields.Integer(
+        string='Previous day', help='Previous day to event date begin')
```
Change string and help according what I said in previous comments.

Commit:
[UPDATE]

Endika commented:
[UPDATE] all

Commit:
[UPDATE] README

Endika commented:
Ok, sorry ;)

Commit:
[UPDATE] "Done tasks shouldn't be moved."

Endika commented:
> Done tasks shouldn't be moved.

Upload

Commit:
[UPDATE]
-You don't need this instruction now
-Don't browse a recordset! self.env.ref already returns the recordset.
-This should be project_task_type_done.id

Endika commented:
OK

> [UPDATE] …
> -You don't need this instruction now
> -Don't browse a recordset! self.env.ref already returns the recordset.
> -This should be project_task_type_done.id

pedrobaeza commented:
:+1:
